File: frontend/src/lib/api/bidang.ts
import { request } from "./client";

export async function createBidang(
  name: string,
  sort_order: number = 0,
  folderIds: string[] = [],
  autoCreateRootFolder: boolean = true
) {
  try {
    if (!name || name.trim() === "") {
      return { success: false, error: "Nama bidang tidak boleh kosong" };
    }
    const res = await request<{
      success: boolean;
      data?: { id: string; name: string; sort_order: number; sortOrder?: number };
      error?: string;
    }>("/bidang", {
      method: "POST",
      body: JSON.stringify({
        name: name.trim(),
        sortOrder: sort_order,
        folderIds,
        autoCreateRootFolder,
      }),
    });
    if (!res.success) throw new Error(res.error || "Gagal menambahkan bidang");
    return { success: true, data: res.data };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

export async function getBidangList() {
  try {
    const res = await request<{ success: boolean; data?: any[]; error?: string }>("/bidang");
    if (!res.success) throw new Error(res.error || "Gagal mengambil data bidang");
    return { success: true, data: res.data ?? [] };
  } catch (error) {
    return { success: false, data: [] as any[], error: error instanceof Error ? error.message : String(error) };
  }
}

export async function getBidangFolders(id: string) {
  try {
    const res = await request<{ success: boolean; data: string[]; error?: string }>(`/bidang/${encodeURIComponent(id)}/folders`);
    if (!res.success) throw new Error(res.error || "Gagal memuat hak akses folder");
    return { success: true, data: res.data ?? [] };
  } catch (error) {
    return { success: false, data: [] as string[], error: error instanceof Error ? error.message : String(error) };
  }
}

export async function updateBidangFolders(id: string, folderIds: string[]) {
  try {
    const res = await request(`/bidang/${encodeURIComponent(id)}/folders`, {
      method: "PUT",
      body: JSON.stringify({ folderIds }),
    });
    if (!res.success) throw new Error(res.error || "Gagal memperbarui hak akses folder");
    return { success: true };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

export async function updateBidang(id: string, name: string, sort_order: number = 0) {
  try {
    if (!name || name.trim() === "") {
      return { success: false, error: "Nama bidang tidak boleh kosong" };
    }
    const res = await request(`/bidang/${encodeURIComponent(id)}`, {
      method: "PATCH",
      body: JSON.stringify({ name: name.trim(), sortOrder: sort_order }),
    });
    if (!res.success) throw new Error(res.error || "Gagal memperbarui bidang");
    return { success: true };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

export async function deleteBidang(id: string) {
  try {
    const res = await request(`/bidang/${encodeURIComponent(id)}`, { method: "DELETE" });
    if (!res.success) throw new Error(res.error || "Gagal menghapus bidang");
    return { success: true };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

export async function reorderBidang(items: { id: string; sort_order: number }[]) {
  try {
    const res = await request("/bidang/reorder", {
      method: "POST",
      body: JSON.stringify({ items }),
    });
    if (!res.success) throw new Error(res.error || "Gagal menyimpan urutan bidang");
    return { success: true };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

File: backend/internal/handler/bidang.go
package handler

import (
	"github.com/gofiber/fiber/v3"

	"github.com/kemenag-baritoutara/betang-kemenag/internal/domain"
	"github.com/kemenag-baritoutara/betang-kemenag/internal/service"
)

// BidangHandler menangani endpoint CRUD bidang.
type BidangHandler struct {
	svc *service.BidangService
}

// NewBidangHandler membuat handler bidang.
func NewBidangHandler(svc *service.BidangService) *BidangHandler {
	return &BidangHandler{svc: svc}
}

// List mengembalikan seluruh bidang beserta jumlah dokumen.
func (h *BidangHandler) List(c fiber.Ctx) error {
	items, err := h.svc.List(c.Context())
	if err != nil {
		return writeFail(c, fiber.StatusInternalServerError, "Gagal memuat bidang.")
	}
	return writeOK(c, items)
}

// Create menambah bidang baru (super admin).
func (h *BidangHandler) Create(c fiber.Ctx) error {
	var req struct {
		Name                 string   `json:"name"`
		FolderIDs            []string `json:"folderIds"`
		AutoCreateRootFolder bool     `json:"autoCreateRootFolder"`
	}
	if err := c.Bind().Body(&req); err != nil {
		return writeFail(c, fiber.StatusBadRequest, "Data permintaan tidak valid.")
	}
	user := currentUser(c)
	created, err := h.svc.Create(c.Context(), req.Name, req.FolderIDs, req.AutoCreateRootFolder, user.ID, user.Email, clientIP(c))
	if err != nil {
		return writeError(c, err)
	}
	return writeOK(c, created)
}

// GetFolders mengambil daftar ID folder root yang diizinkan untuk satu bidang.
func (h *BidangHandler) GetFolders(c fiber.Ctx) error {
	id := c.Params("id")
	ids, err := h.svc.GetAccessibleFolderIDs(c.Context(), id)
	if err != nil {
		return writeFail(c, fiber.StatusInternalServerError, "Gagal memuat hak akses folder.")
	}
	return writeOK(c, ids)
}

// SetFolders memperbarui daftar ID folder root yang diizinkan untuk satu bidang.
func (h *BidangHandler) SetFolders(c fiber.Ctx) error {
	id := c.Params("id")
	var req struct {
		FolderIDs []string `json:"folderIds"`
	}
	if err := c.Bind().Body(&req); err != nil {
		return writeFail(c, fiber.StatusBadRequest, "Data permintaan tidak valid.")
	}
	user := currentUser(c)
	if err := h.svc.SetAccessibleFolders(c.Context(), id, req.FolderIDs, user.Email, clientIP(c)); err != nil {
		return writeError(c, err)
	}
	return writeOK(c, fiber.Map{"success": true})
}

// Update mengubah nama/urutan bidang (super admin).
func (h *BidangHandler) Update(c fiber.Ctx) error {
	var req struct {
		Name      string `json:"name"`
		SortOrder int    `json:"sortOrder"`
	}
	if err := c.Bind().Body(&req); err != nil {
		return writeFail(c, fiber.StatusBadRequest, "Data permintaan tidak valid.")
	}
	id := c.Params("id")
	user := currentUser(c)
	if err := h.svc.Update(c.Context(), id, req.Name, req.SortOrder, user.Email, clientIP(c)); err != nil {
		return writeError(c, err)
	}
	return writeOK(c, fiber.Map{"id": id})
}

// Delete menghapus bidang (super admin).
func (h *BidangHandler) Delete(c fiber.Ctx) error {
	id := c.Params("id")
	user := currentUser(c)
	if err := h.svc.Delete(c.Context(), id, user.Email, clientIP(c)); err != nil {
		return writeError(c, err)
	}
	return writeOK(c, nil)
}

// Reorder memperbarui urutan bidang (super admin).
func (h *BidangHandler) Reorder(c fiber.Ctx) error {
	var req struct {
		Items []struct {
			ID string `json:"id"`
		} `json:"items"`
	}
	if err := c.Bind().Body(&req); err != nil {
		return writeFail(c, fiber.StatusBadRequest, "Data permintaan tidak valid.")
	}
	user := currentUser(c)

	items := make([]domain.Bidang, 0, len(req.Items))
	for i, item := range req.Items {
		items = append(items, domain.Bidang{ID: item.ID, SortOrder: i})
	}
	if err := h.svc.Reorder(c.Context(), items, user.Email, clientIP(c)); err != nil {
		return writeError(c, err)
	}
	return writeOK(c, nil)
}